### ISSUE TITLE
Automatic Village layout

### ISSUE THREAD

**Issue by dgendreau-casana** [G3rste/vsvillage#1]
I gave a few ideas to Kinniken very early in his development of the Millenaire mod. One thing you might want to consider is a way to grow the town from the center outwards and where certain town structures will be placed in relation to the town center without being too rigid about it. At the time, Kinniken had the concept of the town progressing through a number of stages of growth and that certain structures and building upgrades would happen over time. He did this by having a sequence of progressively better prefabs for each structure that become unlocked a certain stages of town growth. One thing I suggested that worked very well was to assign a minimum and maximum radius for each type of structure relative to the town center for each stage of town growth. This provides a way for villages to be randomly generated and grow over time, but still have a reasonable layout. For example, shops and the blacksmith should be near the center of the town while farms should be closer to the outside edges of the town. When the village progresses to a new stage of growth, some structures will be allowed to be replaced, such as a farm that is now to close getting replaced with an inn and then a new farm gets allocated further out. With a framework like this, it also becomes easier to reserve space for and plan out things like roads, walls, watchtowers etc as radial and axial paths. A larger town might also have its farms placed outside the city walls.

**Comment by G3rste** [comment]
But wouldn't this limit players expanding the village themselfes?
I took some inspiration from Tektopia and would like to offer a similar experience in the future, where players could found/ expand villages by recruiting villagers and providing shelter for them.

**Comment by dgendreau-casana** [comment]
Not necessarily. Its just used for helping the village decide where to place buildings. It would make sense for the village to only place new buildings in an open area and not overwrite what the player has built. I know in Millenaire, you can modify the buildings that the village creates and if the village later upgrades that building, it only replaces blocks that were in the original prefab and does not overwrite player edited blocks. I'm sure there is a way to find a reasonable balance between the town growing and the player building within it.

**Comment by Sythelux** [comment]
I also liked the concepts of Millenaire. There are more things like a plot the player can buy that won't be touched by the growing village there is also a progression/quest system and traders between villages. G3rste, If you consider any of those I'd love to help with contributing to this mod as I was about to start my own, but found this project and rather work together.
